k8s/pkg/apis/longhorn/v1beta2/objectstore.go

```diff
@@ -0,0 +1,175 @@
+package v1beta2
+
+import (
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type ObjectStoreState string
+
+const (
+	ObjectStoreStateUnknown  = ObjectStoreState("Unknown")
+	ObjectStoreStateStarting = ObjectStoreState("Starting")
+	ObjectStoreStateRunning  = ObjectStoreState("Running")
+	ObjectStoreStateStopping = ObjectStoreState("Stopping")
+	ObjectStoreStateError    = ObjectStoreState("Error")
+)
+
+type ObjectStoreStatus struct {
+	// The state of the object store as observed by the object store controller.
+	//
+	// The object store implemets a state machine with this, beginning at
+	// "unknown". Once the object store controller detects the new object store
+	// and begins resource creation the state is transformed to "starting". The
+	// object store controller observes the resources and once all resources are
+	// up and ready the state transitions to "running".
+	// The state remains in "running" until the object store is deleted, at
+	// which point the controller will clean up the associated resources.
+	//
+	//  │ object store created
+	//  │
+	//  └──►┌───────────┐
+	//      │ unknown   │
+	//  ┌── └───────────┘
+	//  │
+	//  │ controller creates resources
+	//  │
+	//  └──►┌───────────┐
+	//      │ starting  │
+	//  ┌── └───────────┘
+	//  │
+	//  │ controller detects all resources ready
+	//  │
+	//  └──►┌───────────┐
+	//      │ running   │
+	//  ┌── └───────────┘
+	//  │
+	//  │ object endpoint is marked for deletion
+	//  │
+	//  └──►┌───────────┐
+	//      │ stopping  │
+	//  ┌── └───────────┘
+	//  │
+	//  │ controller waits for dependent resources to disappear before removing
+	//  │ the finalizer, thereby letting the object store be deleted
+	//
+	// +optional
+	// +default="Unknown"
+	State ObjectStoreState `json:"state"`
+
+	// A list of addresses where the S3 store is exposed.
+	//
+	// +optional
+	Endpoints []string `json:"endpoints"`
+}
+
+type ObjectStoreCredentials struct {
+	// An access key. This will be the initial access key for the administrative
+	// account.
+	//
+	// +optional
+	AccessKey string `json:"accessKey"`
+
+	// A secret key. This is the secret key corresponding to the access key above.
+	//
+	// +optional
+	SecretKey string `json:"secretKey"`
+}
+
+type ObjectStoreStorageSpec struct {
```

Comment:
Proposing to add DataLocality. This setting seems pretty popular from our telemetry data

Reply:
I wonder if it wouldn't be a better idea to just make the `.spec.storage` property of type `VolumeSpec` and inherit all properties that a Longhorn volume can have?
We can always ignore or override a property that we're opinionated about or one that doesn't make sense in this context, but that way we could just pass the parameters through to the volume we create without having to copy all the fields by hand. What do you think?

Reply:
I would prefer to specifically select the fields that make sense for ObjectStore instead of inheriting all properties that a Longhorn volume can have. We can always expand/add more field when needed in the future. Also, by doing this, we can add `omitempty` instruction so we don't store empty field to backup target. WDYT?